src/libsyntax/parse/parser.rs

```diff
@@ -1661,6 +1661,8 @@ impl<'a> Parser<'a> {
 
     /// matches '-' lit | lit
     pub fn parse_pat_literal_maybe_minus(&mut self) -> PResult<'a, P<Expr>> {
+        maybe_whole_expr!(self);
```

Comment:
This line permits much more that just negative integer literals, now `PatKind::Lit` and `PatKind::Range` can contain arbitrary expressions.
I'm not sure that later compilation stages are prepared to deal with them, but there's a chance that it will "just work" and passing an arbitrary expression through `expr` matcher will be equivalent to passing it through a constant.

There are two ways to proceed, I think:
- Conservative way: add a check to `ast_validation` ensuring that only paths, literals and literals with minus are accepted in `PatKind::Lit` and `PatKind::Range`.
- Non-conservative way: extend the language and permit passing arbitrary expressions to literal/range patterns through `expr` matchers, make sure it doesn't cause ICEs etc, make @rust-lang/lang aware of this change.

In any case tests for passing arbitrary expressions are required.